doc/CMakeLists.txt

```diff
@@ -346,12 +308,6 @@ add_custom_target(
 )
 add_dependencies(sphinx-html content)
 
-# "breathe", the sphinx plugin that parses XML output from doxygen, has
-# an "everything on everything" dependency issue reported at:
-# https://github.com/michaeljones/breathe/issues/420 In other words
-# changing 1 source file costs the same build time than changing all
-# source files. breathe is fortunately smart enough not to run at all
-# when *nothing* has changed.
 add_custom_target(
```

Comment:
> > It is meant to be used with other plugins such as breathe in order to improve the user experience.
> 
> I'm afraid I don't see yet what is the "user experience" benefit of making doxygen dependent on sphinx. Most people use some higher level target and shouldn't notice any difference. For others who use lower level targets (for troubleshooting or performance reasons), this will make it more complicated or even impossible to run doxygen directly, adding a layer of indirection in an area (the build) that already has too many layers of indirection.

It's not the first plugin to do this. Others such as [exhale](https://exhale.readthedocs.io/en/latest/) do this. Even `breathe` seems to support that (doesn't seem really well designed though), see https://github.com/michaeljones/breathe/blob/master/breathe/process.py

> 
> Considering this PR increases the number of lines there should be some obvious benefits. I've done some testing and it makes no performance difference for incremental builds.

I started this extension in order to simplify the Doxygen build, which now depends on a combination of CMake and an external script. It turned out that the external script is not actually needed (more details below), so it is true that we could keep using CMake for the Doxygen part. The main benefit right now is that we do not need to keep Doxygen input file list duplicate in both Doxyfile and CMakeLists.txt. The extension takes care of parsing that to then keep track of Doxygen changes.
> 
> Speaking of number of lines there was BTW a never implemented suggestion to greatly simplify the timestamp script and maybe even replace it with single recursive copy python call: https://sourceforge.net/p/doxygen/mailman/message/36582383/

The modification time script becomes redundant once input dependency tracking is introduced. With that feature in place, Doxygen is only run when any of the input file is changed. There is an edge case that the script could solve: what if I edit .c/.h, but I don't change any docstring? In theory the generated XML files should be the same, so I could skip updating the XML files given to Sphinx. Well, it turns out that Doxygen renders the timestamp to all XML files, so that case can't be fixed.

> 
> > Doxyfile can be optionally preprocessed so that variables can be inserted
> 
> CMake can pre-process anything, that's routine and does not require a sphinx plugin.

That's true. I added this facility to make transition easier (it's really 5 LOC + extension options).

> 
> > The script to modify timestamps is not used anymore. It was redundant since Doxygen is only run if any of its input files changes and
> 
> Doxygen runs really fast and running it never had any _direct_ performance impact, only an indirect one on breathe.
> 

Detailed above.

> > in case there are no doc string changes, Doxygen still changes rendered timestamps in all XML files...
> 
> I'm very confused by this:
> 
> * The previous line just stated that Doxygen is not run anymore when there no changes

And that's true because there is input dependency tracking.

> * The second line seems to miss the entire purpose of the timestamp script described in its --help and in fact the new script seems to re-implements more or less the same thing. For the record this script made incremental, non-doxygen builds orders of magnitude faster, see #13159 and also #13442

At the time #13159 was introduced CMake did not track Doxygen input file changes. This is not true anymore in both current CMake nor the `doxyrunner` extension. #13442 could have deleted the timestamp script.

Maybe I'm missing something here, but on my tests I see the expected behavior without caring about modification times at all.

Reply:
added reference to breathe 420 issue in the doxyrunner extension.